crates/libm-test/src/mpfloat.rs

```diff
@@ -0,0 +1,389 @@
+
+                impl MpOp for Operation {
+                    type Input = $RustArgs;
+                    type Output = $RustRet;
+
+                    fn new() -> Self {
+                        Self(new_mpfloat::<$fty>(), new_mpfloat::<$fty>(), new_mpfloat::<$fty>())
+                    }
+
+                    fn run(&mut self, input: Self::Input) -> Self::Output {
+                        self.0.assign(input.0);
+                        self.1.assign(input.1);
+                        self.2.assign(input.2);
+                        let ord = self.0.[< $fn_name_normalized _round >](&self.1, &self.2, Nearest);
+                        prep_retval::<Self::Output>(&mut self.0, ord)
+                    }
+                }
+            }
+        }
+    };
+}
+
+libm_macros::for_each_function! {
+    callback: impl_mp_op,
+    skip: [
+        // Most of these need a manual implementation
+        fabs, ceil, copysign, floor, rint, round, trunc,
+        fabsf, ceilf, copysignf, floorf, rintf, roundf, truncf,
+        fmod, fmodf, frexp, frexpf, ilogb, ilogbf, jn, jnf, ldexp, ldexpf,
+        lgamma_r, lgammaf_r, modf, modff, nextafter, nextafterf, pow,powf,
+        remquo, remquof, scalbn, scalbnf, sincos, sincosf,
+    ],
+    fn_extra: match MACRO_FN_NAME {
+        // Remap function names that are different between mpfr and libm
+        expm1 | expm1f => exp_m1,
+        fabs | fabsf => abs,
+        fdim | fdimf => positive_diff,
+        fma | fmaf => mul_add,
+        fmax | fmaxf => max,
+        fmin | fminf => min,
+        lgamma | lgammaf => ln_gamma,
+        log | logf => ln,
+        log1p | log1pf => ln_1p,
+        tgamma | tgammaf => gamma,
+        _ => MACRO_FN_NAME_NORMALIZED
+    }
+}
+
+/// Implement unary functions that don't have a `_round` version
+macro_rules! impl_no_round {
+    // Unary matcher
+    ($($fn_name:ident, $rug_name:ident;)*) => {
+        paste::paste! {
+            // Implement for both f32 and f64
+            $( impl_no_round!{ @inner_unary [< $fn_name f >], (f32,), $rug_name } )*
+            $( impl_no_round!{ @inner_unary $fn_name, (f64,), $rug_name } )*
+        }
+    };
+
+    (@inner_unary $fn_name:ident, ($fty:ty,), $rug_name:ident) => {
+        pub mod $fn_name {
+            use super::*;
+            pub struct Operation(MpFloat);
+
+            impl MpOp for Operation {
+                type Input = ($fty,);
+                type Output = $fty;
+
+                fn new() -> Self {
+                    Self(new_mpfloat::<$fty>())
+                }
+
+                fn run(&mut self, input: Self::Input) -> Self::Output {
+                    self.0.assign(input.0);
+                    self.0.$rug_name();
+                    prep_retval::<Self::Output>(&mut self.0, Ordering::Equal)
+                }
+            }
+        }
+    };
+}
+
+impl_no_round! {
+    fabs, abs_mut;
+    ceil, ceil_mut;
+    floor, floor_mut;
+    rint, round_mut;
```

Comment:
```suggestion
    rint, round_even_mut; // FIXME: respect rounding mode
```
`round` does ties away from zero, `rint` uses the current rounding mode. Rug doesn't seem to have a direct equivalent, but this should be correct until we want to support changing the rounding mode.

Reply:
Thanks, I was wondering what to do here.

Poll: is `expected.validate(actual, ...)` more intuitive or `actual.validate(expected, ...)`? I realize I have them flipped so my output is backwards...